File: app/src/main/java/com/rr/razasypelajes/Helpers/JSONHelper.java
package com.rr.razasypelajes.Helpers;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

public class JSONHelper {
    private static JSONArray getJSON(InputStream is) throws IOException, JSONException {
        Writer writer = new StringWriter();
        char[] buffer = new char[1024];
        Reader reader = new BufferedReader(new InputStreamReader(is));
        int n;
        while ((n = reader.read(buffer)) != -1) writer.write(buffer, 0, n);
        is.close();
        return new JSONArray(writer.toString());
    }

    public static <T extends JSONBuildable> List<T> fromJSON(Class<T> tClass, InputStream is) {
        List<T> list = new ArrayList<>();
        try {
            JSONArray arr = getJSON(is);
            for (int i = 0; i < arr.length(); i++) {
                JSONObject current = arr.getJSONObject(i);
                Constructor<T> constructor = tClass.getDeclaredConstructor(JSONObject.class);
                list.add(constructor.newInstance(current));
            }
        } catch (IOException | JSONException | NoSuchMethodException | InstantiationException |
                IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        return list;
    }
}

File: app/src/main/java/com/rr/razasypelajes/helpers/JSONHelper.kt
package com.rr.razasypelajes.helpers

import org.json.JSONArray
import org.json.JSONObject
import java.io.*
import java.lang.reflect.Constructor

class JSONHelper {
    companion object {
        private fun getJSON(inputStream: InputStream): JSONArray {
            val writer = StringWriter() as Writer
            val buffer = CharArray(1024)
            val reader = BufferedReader(InputStreamReader(inputStream)) as Reader
            var n = reader.read(buffer)
            while (n != -1) {
                writer.write(buffer, 0, n)
                n = reader.read(buffer)
            }
            inputStream.close()
            return JSONArray(writer.toString())
        }

        fun <T : JSONBuildable> fromJSON(tClass: Class<T>, inputStream: InputStream): List<T> {
            val list = ArrayList<T>()
            val arr = getJSON(inputStream)
            for (i in 0 until arr.length()) {
                val current = arr.getJSONObject(i)
                val constructor = tClass.getDeclaredConstructor(JSONObject::class.java) as Constructor<T>
                list.add(constructor.newInstance(current))
            }
            return list
        }

        fun convertJsonArray(jsonArray: JSONArray) : List<String> {
            val list = ArrayList<String>()
            for (i in 0 until jsonArray.length()) list.add(jsonArray[i].toString())
            return list
        }

        fun stringFromJsonArray(jsonArray: JSONArray) : String {
            val sb = StringBuilder()
            for (i in 0 until jsonArray.length() - 1) sb.append(jsonArray[i]).append("\n")
            sb.append(jsonArray[jsonArray.length() - 1])
            return sb.toString()
        }
    }
}
